pageserver/src/layered_repository.rs

```diff
@@ -1162,22 +1162,34 @@ impl LayeredTimeline {
         // check, though. We should also aim at flushing layers that consume
         // a lot of memory and/or aren't receiving much updates anymore.
         let mut disk_consistent_lsn = last_record_lsn;
-        while let Some(oldest_layer) = layers.peek_oldest_open() {
-            // Does this layer need freezing?
+
+        while let Some((oldest_layer, oldest_generation)) = layers.peek_oldest_open() {
             let oldest_pending_lsn = oldest_layer.get_oldest_pending_lsn();
-            let distance = last_valid_lsn.0 - oldest_pending_lsn.0;
-            if !force && distance < OLDEST_INMEM_DISTANCE {
+
+            // Does this layer need freezing?
+            //
+            // Write out all in-memory layers that contain WAL older than OLDEST_INMEM_DISTANCE.
+            // Or if 'force' is true, write out all of them. If we reach a layer with the same
+            // generation number, we know that we have cycled through all layers that were open
+            // when we started. We don't want to process layers inserted after we started, to
+            // avoid getting into an infinite loop trying to process again entries that we
+            // inserted ourselves.
+            let distance:i128 = i128::from(last_record_lsn.0) - i128::from(oldest_pending_lsn.0);
+            if distance < 0
+                || (!force && distance < i128::from(OLDEST_INMEM_DISTANCE))
+                || oldest_generation == current_generation
+            {
                 info!(
-                    "the oldest layer is now {} which is {} bytes behind last_valid_lsn",
-                    oldest_layer.get_seg_tag(),
+                    "the oldest layer is now {} which is {} bytes behind last_record_lsn",
+                    oldest_layer.filename().display(),
                     distance
                 );
                 disk_consistent_lsn = oldest_pending_lsn;
                 break;
             }
 
             // freeze it
-            let (new_historics, new_open) = oldest_layer.freeze(last_valid_lsn, &self)?;
+            let (new_historics, new_open) = oldest_layer.freeze(last_record_lsn, &self)?;
 
             // replace this layer with the new layers that 'freeze' returned
             layers.pop_oldest_open();
```

Comment:
Just to confirm, is the point of `generation` to avoid popping the layers that are inserted earlier in this same loop (L1197)?

Reply:
That's right!